Simplify bound params (Part 1/2)

msmk0 opened the pull request:
- Implement `SingleBoundTrackParameters` without the `SingleTrackParameters` base.
- Implement `SingleCurvilinearTrackParameters` as a sub-class of the bound parameters. Curvilinear parameters are just bound parameters with a fixed surface type. Consequently, they are implemented as a subclass that only differs in the constructors. This enables simpler code for cases that previously had to handle bound and curvilinear track parameters separately.
- Make the bound parameters const-classes, i.e. they can not be modified after construction, by removing the `.set()` functionality. This greatly simplifies the implementation; e.g. we do not have to recompute the surface geometry for curvilinear parameters anymore. This was only used in the `RiddersPropagator`, where this functionality can be replaced by the construction of a new parameters object.
- Remove cached global position and momentum. The `.position()` and `.momentum()` accessors compute the global information directly. In that case of the position the default geometry context is used.

Supersedes #298 and closes #35.

asalzburger commented on Core/include/Acts/Propagator/RiddersPropagator.ipp:
```diff
@@ -175,38 +173,13 @@ Acts::RiddersPropagator<propagator_t>::wiggleDimension(
+
+    // Propagate with updated start parameters
+    BoundParameters tp(startPars.referenceSurface().getSharedPtr(), values,
+                       startPars.covariance());
```
Nice.

asalzburger commented on Examples/Algorithms/Propagation/include/ACTFW/Propagation/PropagationAlgorithm.ipp:
```diff
@@ -169,16 +169,16 @@ ProcessCode PropagationAlgorithm<propagator_t>::execute(
+      Acts::BoundParameters startParameters(surface, std::move(pars),
+                                            std::move(cov));
+      sPosition = startParameters.position(context.geoContext);
       sMomentum = startParameters.momentum();
```
I agree, this makes it really clear: the `Context` object is used when it is needed and not at construction of the `BoundParameters`.

asalzburger commented on Examples/Algorithms/Propagation/include/ACTFW/Propagation/PropagationAlgorithm.ipp:
```diff
@@ -169,16 +169,16 @@ ProcessCode PropagationAlgorithm<propagator_t>::execute(
       Acts::NeutralBoundTrackParameters neutralParameters(
-          context.geoContext, std::move(cov), std::move(pars), surface);
-      sPosition = neutralParameters.position();
+          surface, std::move(pars), std::move(cov));
```
I like the new signature:
bound - to where ? `Surface`, what `Parameters`, how `Covariance`.

Very clear.

asalzburger commented:
Waiting for build problems to be fixed.

msmk0 commented:
@asalzburger I rebased to master, updated to recent changes, and removed the WIP label. After merging #373 the propagation tests should run through. This is now ready to re-reviewed and merged.

asalzburger commented on Tests/IntegrationTests/PropagationTests.hpp:
```diff
@@ -117,9 +117,10 @@ inline Acts::NeutralCurvilinearTrackParameters makeParametersCurvilinearNeutral(
-    double epsDir, double epsMom) {
+    const Acts::SingleBoundTrackParameters<charge_t>& cmp,
+    const Acts::SingleBoundTrackParameters<charge_t>& ref,
+    const Acts::GeometryContext& geoCtx, double epsPos, double epsDir,
```
Usually we have GeometryContext as first argument, but this is a test file, so no problem.

asalzburger replied:
Unrelated to this MR.

asalzburger commented on Tests/UnitTests/Core/EventData/TransformFreeToBoundTests.cpp:
```diff
@@ -0,0 +1,76 @@
+  Vector3D pos = Vector3D::Zero();
+  Vector3D dir = makeDirectionUnitFromPhiTheta(phi, theta);
+  // transform reference position
+  surface->localToGlobal(geoCtx, loc, dir, pos);
```
We might discuss changing this interface at some point ...

asalzburger replied:
Unrelated to this MR.
